### ISSUE TITLE
Should values and errors be dict or view objects?

### ISSUE THREAD

**Issue by cdeil** [iminuit/iminuit#272]
I'm trying out `iminuit` for the first time since your recent changes and noticed this:
```
In [1]:     from iminuit import Minuit
   ...: 
   ...:     def f(x, y, z):
   ...:         return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2
   ...: 
   ...:     m = Minuit(f)
   ...: 
   ...:     m.migrad()  # run optimiser
   ...:     print(m.values)  # {'x': 2,'y': 3,'z': 4}
   ...: 
   ...:     m.hesse()   # run covariance estimator
   ...:     print(m.errors)  # {'x': 1,'y': 1,'z': 1}
   ...: 
   ...: 
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter x does not have initial value. Assume 0.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter x is floating but does not have initial step size. Assume 1.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter y does not have initial value. Assume 0.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter y is floating but does not have initial step size. Assume 1.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter z does not have initial value. Assume 0.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: Parameter z is floating but does not have initial step size. Assume 1.
  
/Users/deil/software/anaconda3/envs/iminuit-dev/bin/ipython:6: InitialParamWarning: errordef is not given. Default to 1.
  
ValueView{ 'x'=2.0000000000047327,
 'y'=3.000000000007099,
 'z'=4.000000000009465
}
ErrorView{ 'x'=1.0000000000000402,
 'y'=0.9999999999999426,
 'z'=1.0000000000001645
}

In [2]: m.errors
Out[2]: <iminuit._libiminuit.ErrorView at 0x11a1216a0>

In [3]: m.errors[0]
Out[3]: 1.0000000000000402
In [8]: json.dumps(m.errors)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-dea7245a2c59> in <module>()
----> 1 json.dumps(m.errors)

~/software/anaconda3/envs/iminuit-dev/lib/python3.6/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    229         cls is None and indent is None and separators is None and
    230         default is None and not sort_keys and not kw):
--> 231         return _default_encoder.encode(obj)
    232     if cls is None:
    233         cls = JSONEncoder

~/software/anaconda3/envs/iminuit-dev/lib/python3.6/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

~/software/anaconda3/envs/iminuit-dev/lib/python3.6/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/software/anaconda3/envs/iminuit-dev/lib/python3.6/json/encoder.py in default(self, o)
    178         """
    179         raise TypeError("Object of type '%s' is not JSON serializable" %
--> 180                         o.__class__.__name__)
    181 
    182     def encode(self, o):

TypeError: Object of type 'ErrorView' is not JSON serializable

```

Before, I think `m.values` and `m.errors` were dicts, no?

I like that one can now use those objects to set.

But some important features are gone:
- Nice repr (easy to add back via `__repr__` that calls `repr(dict(self))`, no?)
- JSON serialisable (can add back via `__dict__` or some other trick?)

Or maybe it's even possible to sub-class [collections.UserDict](https://docs.python.org/3.7/library/collections.html#collections.UserDict) and only add a few methods, instead of trying to implement a dict-like from scratch?

@HDembinski - Thoughts?

**Comment by HDembinski** [comment]
You are right, the views should mimik real dicts so much that other code does not notice the difference. I added methods to that effect, but didn't test this. So what kind of interface does json expect, how can we make this work?

I don't think that views should repr like dicts. repr is supposed to eval back to a valid object, this is not possible here.

**Comment by HDembinski** [comment]
According to the docs of the json module, the only types supported are the basic builtin types. That means a class that imitates the interface of a dict is not enough.

To serialize a view, you can use
`json.dump(m.errors.copy(), file)` or `json.dump(dict(m.errors), file)`

**Comment by cdeil** [comment]
> I don't think that views should repr like dicts. repr is supposed to eval back to a valid object, this is not possible here.

OK to make make the repr different from a dict. But the current repr is super unfriendly to users:
```
`<iminuit._libiminuit.ErrorView at 0x11a1216a0>`
```
Note that Python / ipython / Jupyter by default show the repr of an object when you just put the variable name as a statement, i.e. users will see that all the time.

It's pretty common to put the same `__repr__` and `__str__` on Python objects.
See e.g. https://docs.python.org/3/library/dataclasses.html or `namedtuple` or `OrderedDict` in the Python stdlib.

I'd suggest we do the same. The simplest way is to only define `__repr__` and not `__str__` and then `str` will also show the `repr`.

> To serialize a view, you can use json.dump(m.errors.copy(), file) or json.dump(dict(m.errors), file)

Suggest to rename `copy` to `asdict` or `to_dict`. copy should be copy, and this is returning an object of a different type (dict).

---

With these two little changes, I think it's fine.

@HDembinski - OK? Should I make a PR?

Changing to subclass `collections.UserDict` would still be worth trying, but that's more work.

**Comment by cdeil** [comment]
OK, then can you just also delete the `copy` method, please?

**Comment by HDembinski** [comment]
Both things are implemented in PR #274
